web/frontend/src/components/base/index.svelte

```diff
@@ -0,0 +1,500 @@
+    pressed.add(key);
+    digestInput();
+  };
+
+  const handleKeyUp = (key: Keys) => {
+    pressed.delete(key);
+
+    if (pressed.size > 0) {
+      stopped = false;
+      digestInput();
+    } else {
+      stop();
+    }
+  };
+
+  const handleBaseStraight = async (event: {
+    distance: number;
+    speed: number;
+    direction: number;
+    movementType: MovementTypes;
+  }) => {
+    if (event.movementType === "Continuous") {
+      const linear = { x: 0, y: event.speed * event.direction, z: 0 };
+      const angular = { x: 0, y: 0, z: 0 };
+
+      try {
+        await baseClient.setVelocity(linear, angular);
+      } catch (error) {
+        displayError(error as ServiceError);
+      }
+    } else {
+      try {
+        await baseClient.moveStraight(
+          event.distance,
+          event.speed * event.direction
+        );
+      } catch (error) {
+        displayError(error as ServiceError);
+      }
+    }
+  };
+
+  const baseRun = async () => {
+    if (movementMode === "Spin") {
+      try {
+        await baseClient.spin(
+          angle * (spinType === "Clockwise" ? -1 : 1),
+          spinSpeed
+        );
+      } catch (error) {
+        displayError(error as ServiceError);
+      }
+    } else if (movementMode === "Straight") {
+      handleBaseStraight({
+        movementType,
+        direction: direction === "Forwards" ? 1 : -1,
+        speed,
+        distance: increment,
+      });
+    }
+  };
+
+  const handleViewSelect = (viewMode: View) => {
+    selectedView = viewMode;
+
+    let liveCameras = 0;
+    for (const camera of filteredResources) {
+      if (openCameras[camera.name]) {
+        liveCameras += 1;
+      }
+    }
+    disableViews = liveCameras > 1;
+  };
+
+  const handleTabSelect = (controlMode: Tabs) => {
+    selectedMode = controlMode;
+
+    if (controlMode === "Discrete") {
+      resetDiscreteState();
+    }
+  };
+
+  const handleVisibilityChange = () => {
+    // only issue the stop if there are keys pressed as to not interfere with commands issued by scripts/commands
+    if (document.visibilityState === "hidden" && pressed.size > 0) {
+      pressed.clear();
+      stop();
+    }
+  };
+
+  const handleOnBlur = () => {
+    if (pressed.size <= 0) {
+      stop();
+    }
+  };
+
+  const handleUpdateKeyboardState = (on: boolean) => {
+    isKeyboardActive = on;
+
+    if (isKeyboardActive) {
+      return;
+    }
+
+    if (pressed.size > 0 || !stopped) {
+      stop();
+    }
+  };
+
+  const handleSwitch = (cameraName: string) => {
+    openCameras[cameraName] = !openCameras[cameraName];
+
+    for (const camera of filteredResources) {
+      if (openCameras[camera.name]) {
+        disableRefresh = false;
+        return;
+      }
+    }
+    disableRefresh = true;
+  };
+
+  onMount(() => {
+    window.addEventListener("visibilitychange", handleVisibilityChange);
+
+    // Safety measure for system prompts, etc.
+    window.addEventListener("blur", handleOnBlur);
```

Comment:
probably a pre-existing issue: can we remove this event listener in `onDestroy` as well? It probably wouldn't hurt to also call `handleOnBlur()` in `onDestroy()` too since it represents the user navigating away when this UI is embedded in app